src/jit/lower.h

```diff
@@ -133,30 +133,32 @@ class Lowering : public Phase
 
     //----------------------------------------------------------------------
     // SetRegOptional - sets a bit to indicate to LSRA that register
-    // for a given tree node is optional for codegen purpose.  If no
+    // for a given child node is optional for codegen purpose.  If no
     // register is allocated to such a tree node, its parent node treats
-    // it as a contained memory operand during codegen.
+    // it as a contained memory operand during codegen.  This routine
+    // also updates srcRegOptional count of parent node.
     //
     // Arguments:
-    //    tree    -   GenTree node
+    //    parent  -   Parent GenTree node with 'child' as one of its operands.
+    //    child    -  Child GenTree node that is to marked as reg optional.
     //
     // Returns
     //    None
-    void SetRegOptional(GenTree* tree)
-    {
-        tree->gtLsraInfo.regOptional = true;
-    }
+    void SetRegOptional(GenTree* parent, GenTree* child)
+    {        
+        child->gtLsraInfo.regOptional = true;
 
-    GenTree* PreferredRegOptionalOperand(GenTree* tree);
+        // On xarch, at the most one operand can be honored as
+        // reg optional even if more than one operand of parent
+        // node is marked as reg optional.
+        parent->gtLsraInfo.srcMemOpCount = 1;
+    }   
 
     // ------------------------------------------------------------------
-    // SetRegOptionalBinOp - Indicates which of the operands of a bin-op
-    // register requirement is optional. Xarch instruction set allows
-    // either of op1 or op2 of binary operation (e.g. add, mul etc) to be
-    // a memory operand.  This routine provides info to register allocator
-    // which of its operands optionally require a register.  Lsra might not
-    // allocate a register to RefTypeUse positions of such operands if it
-    // is beneficial. In such a case codegen will treat them as memory
+    // SetRegOptionalBinOp - Marks all the operands of a bin-op that are
+    // permitted to be memory-op as reg optional. Lsra might not allocate
+    // a register to RefTypeUse positions of such operands if it is
+    // beneficial. In such a case codegen will treat them as memory
```

Comment:
This is just a bit confusing I would say "if it is beneficial to leave it in memory" or something.